File: external/js/kyber/src/pairing/gfp6.ts
import GfP2 from './gfp2';
import {
    xiTo2PMinus2Over3,
    xiToPMinus1Over3,
    xiTo2PSquaredMinus2Over3,
    xiToPSquaredMinus1Over3,
} from './constants';
import GfP from './gfp';

/**
 * Group field of size p^6
 * This object acts as an immutable and then any modification will instantiate
 * a new object.
 */
export default class GfP6 {
    private static ZERO = new GfP6();
    private static ONE = new GfP6(GfP2.zero(), GfP2.zero(), GfP2.one());

    /**
     * Get the addition identity for this group field
     * @returns the element
     */
    public static zero(): GfP6 {
        return GfP6.ZERO;
    }

    /**
     * Get the multiplication identity for this group field
     * @returns the element
     */
    public static one(): GfP6 {
        return GfP6.ONE;
    }

    private x: GfP2;
    private y: GfP2;
    private z: GfP2;

    constructor(x?: GfP2, y?: GfP2, z?: GfP2) {
        this.x = x || GfP2.zero();
        this.y = y || GfP2.zero();
        this.z = z || GfP2.zero();
    }

    /**
     * Get the x value of the group field element
     * @returns the x element
     */
    getX(): GfP2 {
        return this.x;
    }

    /**
     * Get the y value of the group field element
     * @returns the y element
     */
    getY(): GfP2 {
        return this.y;
    }

    /**
     * Get the z value of the group field element
     * @returns the z element
     */
    getZ(): GfP2 {
        return this.z;
    }

    /**
     * Check if the element is zero
     * @returns true when zero, false otherwise
     */
    isZero(): boolean {
        return this.x.isZero() && this.y.isZero() && this.z.isZero();
    }

    /**
     * Check if the element is one
     * @returns true when one, false otherwise
     */
    isOne(): boolean {
        return this.x.isZero() && this.y.isZero() && this.z.isOne();
    }

    /**
     * Get the negative of the element
     * @returns the new element
     */
    neg(): GfP6 {
        const x = this.x.negative();
        const y = this.y.negative();
        const z = this.z.negative();
        return new GfP6(x, y, z);
    }

    frobenius(): GfP6 {
        const x = this.x.conjugate().mul(xiTo2PMinus2Over3);
        const y = this.y.conjugate().mul(xiToPMinus1Over3);
        const z = this.z.conjugate();
        return new GfP6(x, y, z);
    }

    frobeniusP2(): GfP6 {
        const x = this.x.mulScalar(new GfP(xiTo2PSquaredMinus2Over3));
        const y = this.y.mulScalar(new GfP(xiToPSquaredMinus1Over3));
        return new GfP6(x, y, this.z);
    }

    /**
     * Add b to the current element
     * @param b the element to add
     * @returns the new element
     */
    add(b: GfP6): GfP6 {
        const x = this.x.add(b.x);
        const y = this.y.add(b.y);
        const z = this.z.add(b.z);
        return new GfP6(x, y, z);
    }

    /**
     * Subtract b to the current element
     * @param b the element to subtract
     * @returns the new element
     */
    sub(b: GfP6): GfP6 {
        const x = this.x.sub(b.x);
        const y = this.y.sub(b.y);
        const z = this.z.sub(b.z);
        return new GfP6(x, y, z);
    }

    /**
     * Multiply the current element by b
     * @param b the element to multiply with
     * @returns the new element
     */
    mul(b: GfP6): GfP6 {
        const v0 = this.z.mul(b.z);
        const v1 = this.y.mul(b.y);
        const v2 = this.x.mul(b.x);

        let t0 = this.x.add(this.y);
        let t1 = b.x.add(b.y);
        let tz = t0.mul(t1);
        tz = tz.sub(v1).sub(v2).mulXi().add(v0);

        t0 = this.y.add(this.z);
        t1 = b.y.add(b.z);
        let ty = t0.mul(t1);
        t0 = v2.mulXi();
        ty = ty.sub(v0).sub(v1).add(t0);

        t0 = this.x.add(this.z);
        t1 = b.x.add(b.z);
        let tx = t0.mul(t1);
        tx = tx.sub(v0).add(v1).sub(v2);

        return new GfP6(tx, ty, tz);
    }

    /**
     * Multiply the current element by a scalar
     * @param b the scalar
     * @returns the new element
     */
    mulScalar(b: GfP2): GfP6 {
        const x = this.x.mul(b);
        const y = this.y.mul(b);
        const z = this.z.mul(b);
        return new GfP6(x, y, z);
    }

    /**
     * Multiply the current element by a GFp element
     * @param b the GFp element
     * @returns the new element
     */
    mulGfP(b: GfP): GfP6 {
        const x = this.x.mulScalar(b);
        const y = this.y.mulScalar(b);
        const z = this.z.mulScalar(b);
        return new GfP6(x, y, z);
    }

    mulTau(): GfP6 {
        const tz = this.x.mulXi();
        
        return new GfP6(this.y, this.z, tz);
    }

    /**
     * Get the square of the current element
     * @returns the new element
     */
    square(): GfP6 {
        const v0 = this.z.square();
        const v1 = this.y.square();
        const v2 = this.x.square();

        const c0 = this.x.add(this.y).square().sub(v1).sub(v2).mulXi().add(v0);
        const c1 = this.y.add(this.z).square().sub(v0).sub(v1).add(v2.mulXi());
        const c2 = this.x.add(this.z).square().sub(v0).add(v1).sub(v2);

        return new GfP6(c2, c1, c0);
    }

    /**
     * Get the inverse of the element
     * @returns the new element
     */
    invert(): GfP6 {
        const A = this.z.square().sub(this.x.mul(this.y).mulXi());
        const B = this.x.square().mulXi().sub(this.y.mul(this.z));
        const C = this.y.square().sub(this.x.mul(this.z));
        const F = C.mul(this.y).mulXi().add(A.mul(this.z)).add(B.mul(this.x).mulXi()).invert();

        return new GfP6(C.mul(F), B.mul(F), A.mul(F));
    }

    /**
     * Check the equality with the other object
     * @param o the other object
     * @returns true when both are equal, false otherwise
     */
    equals(o: any): o is GfP6 {
        return this.x.equals(o.x) && this.y.equals(o.y) && this.z.equals(o.z);
    }

    /**
     * Get the string representation of the element
     * @returns a string representation
     */
    toString(): string {
        return `(${this.x.toString()}, ${this.y.toString()}, ${this.z.toString()})`;
    }
}

File: external/java/src/main/java/ch/epfl/dedis/lib/crypto/bn256/GFp6.java
package ch.epfl.dedis.lib.crypto.bn256;

import java.math.BigInteger;

class GFp6 {
    GFp2 x, y, z;

    GFp6() {
        this.x = new GFp2();
        this.y = new GFp2();
        this.z = new GFp2();
    }

    GFp6(GFp6 a) {
        this.x = new GFp2(a.x);
        this.y = new GFp2(a.y);
        this.z = new GFp2(a.z);
    }

    @Override
    public String toString() {
        return "(" + this.x.toString() + "," + this.y.toString() + "," + this.z.toString() + ")";
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (!(obj instanceof GFp6)) {
            return false;
        }
        GFp6 other = (GFp6)obj;
        return other.x.equals(this.x) && other.y.equals(this.y) && other.z.equals(this.z);
    }

    GFp6 set(GFp6 a) {
        this.x.set(a.x);
        this.y.set(a.y);
        this.z.set(a.z);
        return this;
    }

    GFp6 setZero() {
        this.x.setZero();
        this.y.setZero();
        this.z.setZero();
        return this;
    }

    GFp6 setOne() {
        this.x.setZero();
        this.y.setZero();
        this.z.setOne();
        return this;
    }

    void minimal() {
        this.x.minimal();
        this.y.minimal();
        this.z.minimal();
    }

    boolean isZero() {
        return this.x.isZero() && this.y.isZero() && this.z.isZero();
    }

    boolean isOne() {
        return this.x.isZero() && this.y.isZero() && this.z.isOne();
    }

    GFp6 negative(GFp6 a) {
        this.x.negative(a.x);
        this.y.negative(a.y);
        this.z.negative(a.z);
        return this;
    }

    GFp6 frobenius(GFp6 a) {
        this.x.conjugate(a.x);
        this.y.conjugate(a.y);
        this.z.conjugate(a.z);

        this.x.mul(this.x, Constants.xiTo2PMinus2Over3);
        this.y.mul(this.y, Constants.xiToPMinus1Over3);
        return this;
    }

    GFp6 frobeniusP2(GFp6 a) {
        this.x.mulScalar(a.x, Constants.xiTo2PSquaredMinus2Over3);
        this.y.mulScalar(a.y, Constants.xiToPSquaredMinus1Over3);
        this.z.set(a.z);
        return this;
    }

    GFp6 add(GFp6 a, GFp6 b) {
        this.x.add(a.x, b.x);
        this.y.add(a.y, b.y);
        this.z.add(a.z, b.z);
        return this;
    }

    GFp6 sub(GFp6 a, GFp6 b) {
        this.x.sub(a.x, b.x);
        this.y.sub(a.y, b.y);
        this.z.sub(a.z, b.z);
        return this;
    }

    GFp6 dbl(GFp6 a) {
        this.x.dbl(a.x);
        this.y.dbl(a.y);
        this.z.dbl(a.z);
        return this;
    }

    GFp6 mul(GFp6 a, GFp6 b) {
        GFp2 v0 = GFp2Pool.getInstance().get();
        v0.mul(a.z, b.z);
        GFp2 v1 = GFp2Pool.getInstance().get();
        v1.mul(a.y, b.y);
        GFp2 v2 = GFp2Pool.getInstance().get();
        v2.mul(a.x, b.x);

        GFp2 t0 = GFp2Pool.getInstance().get();
        t0.add(a.x, a.y);
        GFp2 t1 = GFp2Pool.getInstance().get();
        t1.add(b.x, b.y);
        GFp2 tz = GFp2Pool.getInstance().get();
        tz.mul(t0, t1);

        tz.sub(tz, v1);
        tz.sub(tz, v2);
        tz.mulXi(tz);
        tz.add(tz, v0);

        t0.add(a.y, a.z);
        t1.add(b.y, b.z);
        GFp2 ty = GFp2Pool.getInstance().get();
        ty.mul(t0, t1);
        ty.sub(ty, v0);
        ty.sub(ty, v1);
        t0.mulXi(v2);
        ty.add(ty, t0);

        t0.add(a.x, a.z);
        t1.add(b.x, b.z);
        GFp2 tx = GFp2Pool.getInstance().get();
        tx.mul(t0, t1);
        tx.sub(tx, v0);
        tx.add(tx, v1);
        tx.sub(tx, v2);

        this.x.set(tx);
        this.y.set(ty);
        this.z.set(tz);

        GFp2Pool.getInstance().put(v0, v1, v2, t0, t1, tz, ty, tx);

        return this;
    }

    GFp6 mulScalar(GFp6 a, GFp2 b) {
        this.x.mul(a.x, b);
        this.y.mul(a.y, b);
        this.z.mul(a.z, b);
        return this;
    }

    GFp6 mulGFP(GFp6 a, BigInteger b) {
        this.x.mulScalar(a.x, b);
        this.y.mulScalar(a.y, b);
        this.z.mulScalar(a.z, b);
        return this;
    }

    GFp6 mulTau(GFp6 a) {
        GFp2 tz = GFp2Pool.getInstance().get();
        tz.mulXi(a.x);
        GFp2 ty = GFp2Pool.getInstance().get();
        ty.set(a.y);

        this.y.set(a.z);
        this.x.set(ty);
        this.z.set(tz);

        GFp2Pool.getInstance().put(ty, tz);
        return this;
    }

    GFp6 square(GFp6 a) {
        GFp2 v0 = GFp2Pool.getInstance().get().square(a.z);
        GFp2 v1 = GFp2Pool.getInstance().get().square(a.y);
        GFp2 v2 = GFp2Pool.getInstance().get().square(a.x);

        GFp2 c0 = GFp2Pool.getInstance().get().add(a.x, a.y);
        c0.square(c0);
        c0.sub(c0, v1);
        c0.sub(c0, v2);
        c0.mulXi(c0);
        c0.add(c0, v0);

        GFp2 c1 = GFp2Pool.getInstance().get().add(a.y, a.z);
        c1.square(c1);
        c1.sub(c1, v0);
        c1.sub(c1, v1);
        GFp2 xiV2 = GFp2Pool.getInstance().get().mulXi(v2);
        c1.add(c1, xiV2);

        GFp2 c2 =GFp2Pool.getInstance().get().add(a.x, a.z);
        c2.square(c2);
        c2.sub(c2, v0);
        c2.add(c2, v1);
        c2.sub(c2, v2);

        this.x.set(c2);
        this.y.set(c1);
        this.z.set(c0);

        GFp2Pool.getInstance().put(v0, v1, v2, xiV2, c0, c1, c2);

        return this;
    }

    GFp6 invert(GFp6 a) {
        GFp2 t1 = GFp2Pool.getInstance().get();

        GFp2 A = GFp2Pool.getInstance().get();
        A.square(a.z);
        t1.mul(a.x, a.y);
        t1.mulXi(t1);
        A.sub(A, t1);

        GFp2 B = GFp2Pool.getInstance().get();
        B.square(a.x);
        B.mulXi(B);
        t1.mul(a.y, a.z);
        B.sub(B, t1);

        GFp2 C = GFp2Pool.getInstance().get();
        C.square(a.y);
        t1.mul(a.x, a.z);
        C.sub(C, t1);

        GFp2 F = GFp2Pool.getInstance().get();
        F.mul(C, a.y);
        F.mulXi(F);
        t1.mul(A, a.z);
        F.add(F, t1);
        t1.mul(B, a.x);
        t1.mulXi(t1);
        F.add(F, t1);

        F.invert(F);

        this.x.mul(C, F);
        this.y.mul(B, F);
        this.z.mul(A, F);

        GFp2Pool.getInstance().put(t1, A, B, C, F);

        return this;
    }
}
